### PR TITLE
Migrate dotnet new Cli, dotnet-new3, *.IntegrationTests, *.UnitTests, template_feed to SDK

### PR DESCRIPTION
**Pull request by GangWang01**
Issue - https://github.com/dotnet/templating/issues/4353

Migrate the following parts from dotnet/templating to dotnet/sdk.

- Microsoft.TemplateEngine.Cli
- dotnet-new3
- *.UnitTests, *.IntegrationTests.
- template_feed

### CODE REVIEW

File: eng/Versions.props [hunk 1]
```diff
@@ -124,6 +124,8 @@
     <!-- Dependencies from https://github.com/dotnet/templating -->
     <MicrosoftTemplateEngineCliPackageVersion>7.0.100-preview.7.22358.1</MicrosoftTemplateEngineCliPackageVersion>
```
Comment (vlada-shubina) [inline eng/Versions.props]: after cut-off this is no longer needed

File: src/template_feed/Microsoft.DotNet.Common.ItemTemplates/Microsoft.DotNet.Common.ItemTemplates.csproj [hunk 2]
```diff
@@ -0,0 +1,28 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
Comment (vlada-shubina) [inline src/template_feed/Microsoft.DotNet.Common.ItemTemplates/Microsoft.DotNet.Common.ItemTemplates.csproj:1]: should it be in the root folder instead? e.g. `/template_feed` instead of `/src/template_feed`?

File: eng/Versions.props [hunk 3]
```diff
@@ -124,6 +124,8 @@
     <!-- Dependencies from https://github.com/dotnet/templating -->
     <MicrosoftTemplateEngineCliPackageVersion>7.0.100-preview.7.22361.3</MicrosoftTemplateEngineCliPackageVersion>
```
Comment (vlada-shubina) [inline eng/Versions.props:125]: this needs to be removed - we no longer need the package from dotnet/templating

File: src/template_feed/Microsoft.DotNet.Common.ItemTemplates/Microsoft.DotNet.Common.ItemTemplates.csproj (hunk 2 above)
Comment (GangWang01) [reply src/template_feed/Microsoft.DotNet.Common.ItemTemplates/Microsoft.DotNet.Common.ItemTemplates.csproj:1]: As we discussed before, it's /src/template_feed. Putting it in root folder makes more sense since it's not part of SDK. I will make this change.

**Comment by vlada-shubina** [comment]
@dotnet/dotnet-cli this PR moves `Microsoft.TemplateEngine.Cli`, templates and related tests to dotnet/sdk. 
Could you please review the changes to the scripts and shared projects? 

Thank you.

**Review by dsplaisted — APPROVED** [review]
I didn't review the yml changes, but LGTM.

File: eng/build.yml [hunk 4]
```diff
@@ -320,3 +321,168 @@ jobs:
...
+                -configuration $(_BuildConfig)
+                -prepareMachine
+                -ci
+                -restore
+                $(_Test)
+                -projects $(Build.SourcesDirectory)\src\Tests\Microsoft.TemplateEngine.Cli.UnitTests\Microsoft.TemplateEngine.Cli.UnitTests.csproj
+                /bl:$(Build.SourcesDirectory)\artifacts\log\$(_BuildConfig)\TemplateEngine.Cli.UnitTests.binlog
+                $(_InternalRuntimeDownloadArgs)
+          displayName: Run TemplateEngine Cli Unit Tests
+          condition: succeededOrFailed()
```
Comment (vlada-shubina) [inline eng/build.yml:413]: nit: this project may be run together with other jobs. We can move it in different PR.

File: eng/Versions.props [hunk 5]
```diff
@@ -123,8 +123,9 @@
   </PropertyGroup>
   <PropertyGroup>
     <!-- Dependencies from https://github.com/dotnet/templating -->
-    <MicrosoftTemplateEngineCliPackageVersion>7.0.100-rc.1.22379.2</MicrosoftTemplateEngineCliPackageVersion>
     <MicrosoftTemplateEngineAbstractionsPackageVersion>7.0.100-rc.1.22379.2</MicrosoftTemplateEngineAbstractionsPackageVersion>
+    <MicrosoftTemplateEngineEdgePackageVersion>7.0.100-rc.1.22379.2</MicrosoftTemplateEngineEdgePackageVersion>
```
Comment (vlada-shubina) [inline eng/Versions.props:127]: This change requires dependencies to be changed as well, otherwise it won't be updated automatically. Since all the packages are updated at the same time I suggest we stick to single dependency.

```
  <PropertyGroup>
    <!-- Dependencies from https://github.com/dotnet/templating -->
    <MicrosoftTemplateEngineAbstractionsPackageVersion>7.0.100-rc.1.22401.3</MicrosoftTemplateEngineAbstractionsPackageVersion>
    <MicrosoftTemplateEngineOrchestratorRunnableProjectsPackageVersion>$(MicrosoftTemplateEngineAbstractionsPackageVersion)</MicrosoftTemplateEngineOrchestratorRunnableProjectsPackageVersion>
    <MicrosoftTemplateEngineUtilsPackageVersion>$(MicrosoftTemplateEngineAbstractionsPackageVersion)</MicrosoftTemplateEngineUtilsPackageVersion>
and so on
  </PropertyGroup>

```

File: eng/build.yml (hunk 4 above)
Comment (vlada-shubina) [reply eng/build.yml:413]: Unit tests should be run prior to integration tests

**Comment by richlander** [comment]
Is something wrong with the history on this change? Seems like a lot of commits.

**Comment by vlada-shubina** [comment]
> Is something wrong with the history on this change? Seems like a lot of commits.

We move part of dotnet\templating here, preserving all the commits, so having so many commits is ok.

**Review by vlada-shubina — APPROVED** [review]

**Comment by mmitche** [comment]
@vlada-shubina How does this affect dependency flow for .NET 7? Does this mean that templating isn't part of the build graph?

**Comment by vlada-shubina** [comment]
@mmitche we just moved CLI part for now, so templating is still part of the graph but churn should decrease.
We also removed couple of subscriptions in-between sdk and templating. We are considering to remove dep on runtime for templating in .NET 8 as well.

**Comment by mmitche** [comment]
Okay, so it's basically in its existing spot in the graph, potentially without a runtime dependency. Is it likely that templating will move out of the graph for .NET 8?

**Comment by mmitche** [comment]
@vlada-shubina I have some other questions
- Why is there a separate set of build legs for the template engine?
- Why does it appear in the official build too?
- Why was the `agentOS` parameter to for AzDO templates used to add these? This seems pretty off.

I would expect that these new projects would simply get built alongside the rest of the SDK projects.

**Comment by dsplaisted** [comment]
I believe the separate legs are there just for running the tests separately to try to avoid slowing down overall PR times.  I think as far as the product build goes it should all be the same build.

**Comment by mmitche** [comment]
It seems odd to use "agentOS" for that...seems a bit hacky? Also, the official build is showing the templateengine leg.

**Comment by vlada-shubina** [comment]
> Okay, so it's basically in its existing spot in the graph, potentially without a runtime dependency. Is it likely that templating will move out of the graph for .NET 8?

The repo will remain on its current place, however we expect less commits/updates. Removing runtime dependency is very likely.

**Comment by vlada-shubina** [comment]
> @vlada-shubina I have some other questions
> 
> - Why is there a separate set of build legs for the template engine?
> 
> - Why does it appear in the official build too?
> 
> - Why was the `agentOS` parameter to for AzDO templates used to add these? This seems pretty off.
> 
> 
> 
> I would expect that these new projects would simply get built alongside the rest of the SDK projects.

Template engine and templates tests are running ~1h, therefore we created separate jobs for them so they run in parallel with other jobs. Those tests are not run in other SDK jobs. If we combine all the tests in current single SDK job, the job will likely take about 2 hours and we wanted to avoid that.

We still want to run them in official build to minimize the risk something is broken during official build.

I will check on 'agentOS' tomorrow. @GangWang01 can you please comment on why it was used here?

**Comment by GangWang01** [comment]
> > @vlada-shubina I have some other questions
> > 
> > * Why is there a separate set of build legs for the template engine?
> > * Why does it appear in the official build too?
> > * Why was the `agentOS` parameter to for AzDO templates used to add these? This seems pretty off.
> > 
> > I would expect that these new projects would simply get built alongside the rest of the SDK projects.
> 
> Template engine and templates tests are running ~1h, therefore we created separate jobs for them so they run in parallel with other jobs. Those tests are not run in other SDK jobs. If we combine all the tests in current single SDK job, the job will likely take about 2 hours and we wanted to avoid that.
> 
> We still want to run them in official build to minimize the risk something is broken during official build.
> 
> I will check on 'agentOS' tomorrow. @GangWang01 can you please comment on why it was used here?

As the parameter definition, `agentOS` is not just Agent OS identifier and also used as job name. Following existing SDK build job that use agentOS determining if running AoT jobs, agentOS with “TemplateEngine” distinguishes template engine jobs from SDK jobs.
If using parameter agentOS is not reasonable, a separate build yml file for template engine can avoid using agentOS.
